lexers/asciidoc.lua

```diff
@@ -0,0 +1,110 @@
+-- Copyright 2025 Matěj Cepl. See LICENSE.
+-- Asciidoc LPeg lexer.
+
+local lexer = lexer
+local P, S, B = lpeg.P, lpeg.S, lpeg.B
+
+local lex = lexer.new(...)
+
+-- Distinguish between horizontal and vertical space so html start rule has a chance to match.
+lex:modify_rule('whitespace', lex:tag(lexer.WHITESPACE, S(' \t')^1 + S('\r\n')^1))
+
+-- Operators.
+lex:add_rule('operator', lex:tag(lexer.OPERATOR, S("*:^'`|~+<=>")))
+
+-- Keywords
+lex:add_rule('keyword', lex:tag(lexer.KEYWORD, lex:word_match(lexer.KEYWORD)))
+lex:set_word_list(lexer.KEYWORD, {"NOTE", "IMPORTANT", "WARNING", "TIP", "CAUTION", "TESTME"})
+
+-- Block elements.
+local function h(n)
+	return lex:tag(string.format('%s.h%s', lexer.HEADING, n),
+		lexer.to_eol(lexer.starts_line(string.rep('#', n))))
+end
+lex:add_rule('header', h(6) + h(5) + h(4) + h(3) + h(2) + h(1))
+
+lex:add_rule('hr',
```

Comment:
This rule will never match for `*`, which is highlighted as an operator.

Reply:
You are right, this is mostly based on something I found somewhere on the Web. And yes, you are also right that `operator` rule has to go. What does operator actually mean in the context of a document? Will remove.

Reply:
Rules can have limited backtracking, like `lex:add_rule('method', -lpeg.B('.') * lexer.word * #P('('))`, but it's often not possible with more complicated contexts because the lexer can lose track of state. You can try making use of the `lexer.line_state` table, but it's hard to get it right.